### PR TITLE
feat: add types

### PR DESCRIPTION
**Pull request by hugomrdias**
Changes:

0c911f0 (Hugo Dias, 9 seconds ago)
   fix: fix possible undefined datastores

   - split datastore instantiation from opening
   -  add it-pushable types
   - add just-range types
   - add proper-lockfile types
   - fix lock.close we werent waiting for the promise

1fba61c (Hugo Dias, 24 hours ago)
   feat: base types

   - adds types
   - removes fake async
   - fixes tests for async errors that now are sync
   - adds local types for merge-options

### CODE REVIEW

File: src/types.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+  /**
+   * Map for backends and implementation reference.
+   * - `root` (defaults to `datastore-fs` in Node.js and `datastore-level` in the browser)
+   * - `blocks` (defaults to `datastore-fs` in Node.js and `datastore-level` in the browser)
+   * - `keys` (defaults to `datastore-fs` in Node.js and `datastore-level` in the browser)
+   * - `datastore` (defaults to `datastore-level`)
+   * - `pins` (defaults to `datastore-level`)
+   */
+  storageBackends?: Partial<Record<Backends, DatastoreFactory>>
```
Comment (achingbrain) [inline src/types.ts:29]:
```suggestion
  storageBackends?: Partial<Record<Backends, { new(...args: any[]): Datastore }>>
```

File: src/types.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+  /**
+   * Map for backends and implementation reference.
+   * - `root` (defaults to `datastore-fs` in Node.js and `datastore-level` in the browser)
+   * - `blocks` (defaults to `datastore-fs` in Node.js and `datastore-level` in the browser)
+   * - `keys` (defaults to `datastore-fs` in Node.js and `datastore-level` in the browser)
+   * - `datastore` (defaults to `datastore-level`)
+   * - `pins` (defaults to `datastore-level`)
+   */
+  storageBackends: Record<Backends, DatastoreFactory>
```
Comment (achingbrain) [inline src/types.ts:59]:
```suggestion
  storageBackends: Record<Backends, { new(...args: any[]): Datastore }>
```

File: src/types.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
+import type { DatastoreFactory } from 'interface-datastore'
```
Comment (achingbrain) [inline src/types.ts:1]:
```suggestion
import type { Datastore } from 'interface-datastore'
```

**Review by achingbrain — CHANGES_REQUESTED** [review]

File: types/merge-options/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,2 @@
+declare function mergeOptions<T1, T2> (arg1: T1, arg: T2): T1 & T2
+export = mergeOptions
```
Comment (achingbrain) [inline types/merge-options/index.d.ts:2]: `merge-options` has types in v3, is this necessary?

File: types/proper-lockfile/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
+declare function lock (file: string, options?: Options): Promise<release>
```
Comment (achingbrain) [inline types/proper-lockfile/index.d.ts:1]: The smoking gun:

image

Can these types not be added to `proper-lockfile`?

**Comment by hugomrdias** [comment]
ready to go

**Review by achingbrain — APPROVED** [review]

**Pull request merged**